### ISSUE TITLE
Isagen and CG extensibility improvements

### ISSUE THREAD

**Issue by ldcouto** [overturetool/overture#428]
This issue is meant to track various extensibility improvements done to the codegen  component as part of the trial implementation of Isabelle code generation.

Relevant commits made before opening the issue (oops) include:

* 6b41d7657f761c3d15fd1a08d71ab30b910276a3
* 1cd4ddea746f73dc230a9db8b58f8837a54b0027
* 4932fdd243ab74f8b0d4c973e9e98fbca663653d
*  a6b682f6d8e43b95962704593f1462f050b94335
* 156a9bd562f2a4ce7aa11a8f878147d01a33a6dc

**Comment by ldcouto** [comment]
Isa CG is done so I'm closing this.

**Comment by ldcouto** [comment]
More improvements!

Support for  adding new template folders: 5fe57f80e8 and 675594b621

**Comment by peterwvj** [comment]
Also we have discovered some problems with the IR constructor which imposes several limitations on certain expressions. For example, the IR constructor did now allow the `forall` expression to be constructed outside an imperative context such that the `forall` expression can be replaced with statements which perform the computation of the `forall`. Now that the code generation platform is used for much more than the Java code generator these limitations has to go out of the IR constructor. So what I did was to introduce as mechanism to validate the VDM AST prior to code generating it to Java, which allow us to remove the limitations in the IR constructor.
